### PR TITLE
ENH: partial string indexing on non-monotonic PeriodIndex

### PR DESCRIPTION
**Pull request by jbrockmendel**
Looks like the DatetimeIndex analogue of this was done in #2437 (https://github.com/pandas-dev/pandas/commit/3a173f19b308d2fb7d8e96dfc57f2e5ecd046138)

This does _not_ yet support two-sides slicing e.g. `ser["2015":"2016"]`, but im hoping to get that handled before long.

Partial overlap with #31058, nothing should actually conflict.

### CODE REVIEW

**Comment by WillAyd** [comment]
Is this a bug fix or a new feature? Can you add a whatsnew?

**Comment by jbrockmendel** [comment]
> Is this a bug fix or a new feature? Can you add a whatsnew?

I think it counts as a feature, but there are a lot of inconsistencies in the surrounding code, so I wouldn't argue the point.  Just added whatsnew.

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -13,6 +13,24 @@ including other versions of pandas.
 Enhancements
 ~~~~~~~~~~~~
 
+.. _whatsnew_110.period_index_partial_string_slicing:
+
+Nonmonotonic PeriodIndex Partial String Slicing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+:class:`PeriodIndex` now supports partial string slicing for non-monotonic indexes, mirroring :class:`DatetimeIndex` behavior (:issue:`31096`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: can you update the docs: https://pandas.io/docs/user_guide/timeseries.html#partial-string-indexing

File: doc/source/whatsnew/v1.1.0.rst [hunk 2]
```diff
@@ -13,6 +13,24 @@ including other versions of pandas.
...
+
+For example:
+
+.. ipython:: python
+
+   dti = pd.date_range("2014-01-01", periods=30, freq="30D")
+   pi = dti.to_period("D")
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: can you show ser, maybe break this into 2 to make it more readable

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you rebase

File: doc/source/user_guide/timeseries.rst [hunk 3]
```diff
@@ -1981,6 +1981,9 @@ As with ``DatetimeIndex``, the endpoints will be included in the result. The exa
 
    dfp['2013-01-01 10H':'2013-01-01 11H']
 
+As of version 1.1.0, this works with non-monotonic ``PeriodIndex`` indexes.
```
Comment (jreback) [inline doc/source/user_guide/timeseries.rst]: ok, should be versionadded directive. (ad put at the beginning of the slicing with period strings section); thought we didn't have *any* docs on this (but seems we do)

File: doc/source/whatsnew/v1.1.0.rst [hunk 4]
```diff
@@ -13,6 +13,26 @@ including other versions of pandas.
...
+   ser_monotonic = pd.Series(np.arange(30), index=pi)
+   shuffler = list(range(0, 30, 2)) + list(range(1, 31, 2))
+   ser = ser_monotonic[shuffler]
+   ser
+
+   ser["2014"]
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: can you make a separate ipythonblock for these last 2

**Review by jreback — APPROVED** [review]
doc comments, otherwise lgtm.

**Comment by jreback** [comment]
one more rebase

**Comment by jbrockmendel** [comment]
rebased+green

**Comment by WillAyd** [comment]
Thanks @jbrockmendel

**Pull request merged**